# Lasso, Ridge and Elastic Net Regularization

 Housing price prediction using the Boston Housing Dataset with regularization techniques that reduce overfitting by controlling model complexity and shrinking feature coefficients.

In [16]:
import numpy as np
import pandas as pd
from tensorflow.keras.datasets import boston_housing
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import (RidgeCV, LassoCV, ElasticNetCV)
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
(X_train, y_train), (X_test, y_test) = boston_housing.load_data()

57026/57026 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step


In [17]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [18]:
def evaluate_model(name, y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    r2 = r2_score(y_true, y_pred)

    print(f"\n{name}")
    print(f"MSE: {mse:.4f}")
    print(f"R² Score: {r2:.4f}")

In [19]:
def ridge_regression(X_train, X_test, y_train, y_test):
    alphas = np.logspace(-3, 3, 50)

    model = RidgeCV(alphas=alphas, cv=5)
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    print(f"\nRidge Best Alpha: {model.alpha_}")
    evaluate_model("Ridge Regression (Tuned)", y_test, y_pred)

In [20]:
def lasso_regression(X_train, X_test, y_train, y_test):
    model = LassoCV(
        alphas=np.logspace(-3, 1, 50),
        cv=5,
        max_iter=10000
    )
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    print(f"\nLasso Best Alpha: {model.alpha_}")
    evaluate_model("Lasso Regression (Tuned)", y_test, y_pred)

In [21]:
def elastic_net_regression(X_train, X_test, y_train, y_test):
    model = ElasticNetCV(
        l1_ratio=[0.1, 0.3, 0.5, 0.7, 0.9],
        alphas=np.logspace(-3, 2, 50),
        cv=5,
        max_iter=10000
    )
    model.fit(X_train, y_train)

    y_pred = model.predict(X_test)
    print(f"\nElastic Net Best Alpha: {model.alpha_}")
    print(f"Elastic Net Best l1_ratio: {model.l1_ratio_}")
    evaluate_model("Elastic Net Regression (Tuned)", y_test, y_pred)

In [22]:
ridge_regression(X_train, X_test, y_train, y_test)
lasso_regression(X_train, X_test, y_train, y_test)
elastic_net_regression(X_train, X_test, y_train, y_test)


Ridge Best Alpha: 8.286427728546842

Ridge Regression (Tuned)
MSE: 22.6046
R² Score: 0.7285

Lasso Best Alpha: 0.01151395399326447

Lasso Regression (Tuned)
MSE: 23.0105
R² Score: 0.7236

Elastic Net Best Alpha: 0.013257113655901095
Elastic Net Best l1_ratio: 0.1

Elastic Net Regression (Tuned)
MSE: 22.7969
R² Score: 0.7261


Ridge performs best with the lowest error (MSE ≈ 22.60) and highest explained variance (R² ≈ 0.7285), indicating that shrinking coefficients without eliminating them works well for this dataset. Lasso shows slightly lower performance, suggesting that removing features is less effective here, while Elastic Net gives balanced results close to Ridge but not better.